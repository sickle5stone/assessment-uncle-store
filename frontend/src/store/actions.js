import { Intent, Toaster } from "@blueprintjs/core";

import axios from "axios";

const handleError = (error) => {
  const toaster = Toaster.create();
  toaster.show({ intent: Intent.DANGER, message: error?.message || "Error" });
};

const handleSuccess = (message) => {
  const toaster = Toaster.create();
  toaster.show({ intent: Intent.SUCCESS, message });
};

const backendApiHost = "https://uncle-store-acs4ewqggq-uc.a.run.app";

const actions = (store) => ({
  /**
    Retrieves the list of items from database
  */
  getList: (state) => {
    const currentState = store.getState();
    store.setState({
      ...currentState,
      loading: true,
    });

    axios
      .get(`${backendApiHost}/listItems`)
      .then((response) => {
        store.setState({
          items: response.data,
          loading: false,
        });
      })
      .catch((e) => {
        store.setState({
          loading: false,
        });
        handleError();
      });
  },

  /**
    Adds Item to database via API request
    Payload: { name: string, price: number }
   */
  addItem: (state, payload) => {
    const currentState = store.getState();
    store.setState({
      ...currentState,
      loading: true,
    });

    axios
      .post(`${backendApiHost}/addItem`, { ...payload.input })
      .then((response) => {
        store.setState({
          loading: false,
        });

        handleSuccess(`${response.data.name || "item"} added!`);
      })
      .catch((e) => {
        store.setState({
          loading: false,
        });
        handleError();
      });
  },

  /**
    Updates item in database
   */
  updateItem: (state, payload) => {
    const currentState = store.getState();
    store.setState({
      ...currentState,
      loading: true,
    });

    axios
      .post(`${backendApiHost}/updateItem`, { ...payload.editedItem })
      .then((response) => {
        store.setState({
          loading: false,
        });

        handleSuccess(`${response.data.name || "item"} updated!`);
      })
      .catch((e) => {
        store.setState({
          loading: false,
        });
        handleError();
      });
  },

  /**
    Deletes item in database
   */
  deleteItem: (state, payload) => {
    const currentState = store.getState();
    store.setState({
      ...currentState,
      loading: true,
    });

    axios
      .delete(`${backendApiHost}/deleteItem`, { data: payload })
      .then((response) => {
        store.setState({
          loading: false,
        });

        handleSuccess(`${response.data.name || "item"} deleted!`);
      })
      .catch((e) => {
        store.setState({
          loading: false,
        });
        handleError();
      });
  },
});
export default actions;
