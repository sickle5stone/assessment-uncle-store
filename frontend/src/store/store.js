import createStore from "redux-zero";

const initialState = { items: [] };
const store = createStore(initialState);

export default store;