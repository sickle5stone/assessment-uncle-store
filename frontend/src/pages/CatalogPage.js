import {
  Alignment,
  Button,
  Classes,
  Dialog,
  Navbar,
  NonIdealState,
  Spinner,
} from "@blueprintjs/core";

import { Card } from "../component/Card";
import { ItemDialog } from "../component/ItemDialog";
import React from "react";
import actions from "../store/actions.js";
import { connect } from "redux-zero/react";
import css from "./CatalogPage.module.css";

class CatalogPage extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isOpen: false,
      isAdminOpen: false,
      adminMode: true,
      input: {
        name: "",
        price: 0,
      },
    };
  }

  componentDidMount() {
    const { getList } = this.props;
    getList();
  }

  renderCards = () => {
    const { items } = this.props;
    const { adminMode } = this.state;

    return items?.map((item) => {
      return (
        <Card
          key={item.id}
          data={item}
          className={css.card}
          canEdit={adminMode}
          handleUpdateItem={this.handleUpdateItem}
          handleDeleteItem={this.handleDeleteItem}
        />
      );
    });
  };

  renderAdminDialog = () => {
    const { isAdminOpen } = this.state;

    return (
      <Dialog isOpen={isAdminOpen} className={Classes.DIALOG}>
        <div className={Classes.DIALOG_BODY}>
          <p>
            This is to simulate a login screen, SSO / 3rd party SAML provider
            login can be done here
          </p>
          <Button onClick={() => this.handleLogin()} intent="Success" large>
            Login!
          </Button>
          <Button onClick={() => this.handleLogout()} intent="danger" large>
            Logout
          </Button>
        </div>
      </Dialog>
    );
  };

  renderDialog = () => {
    const { isOpen, input } = this.state;
    return (
      <ItemDialog
        handleOpenDialog={this.handleOpenDialog}
        handleAddItem={this.handleAddItem}
        handleOnChange={this.handleOnChange}
        handlePriceChange={this.handlePriceChange}
        handleClose={this.handleOpenDialog}
        isOpen={isOpen}
        input={input}
      />
    );
  };

  handleDeleteItem = (itemToDelete) => {
    this.setState({ loading: true });
    const { deleteItem, getList } = this.props;

    deleteItem({ itemToDelete });

    setTimeout(() => {
      getList();
      this.setState({ loading: false });
    }, 1000);
  };

  handleUpdateItem = (editedItem) => {
    this.setState({ loading: true });
    const { updateItem, getList } = this.props;

    updateItem({ editedItem });

    setTimeout(() => {
      getList();
      this.setState({ loading: false });
    }, 1000);
  };

  handleLogin = () => {
    this.setState({ isAdminOpen: false, adminMode: true });
  };

  handleLogout = () => {
    this.setState({ isAdminOpen: false, adminMode: false });
  };

  handleOnChange = (e) => {
    const { input } = this.state;
    this.setState({ input: { ...input, [e.target.id]: e.target.value } });
  };

  handlePriceChange = (num, numString) => {
    const { input } = this.state;
    this.setState({ input: { ...input, price: numString } });
  };

  handleAddItem = () => {
    this.setState({ loading: true });

    const { addItem, getList } = this.props;
    const { input } = this.state;
    addItem({ input });

    // Close dialog and refresh page
    this.handleOpenDialog();
    setTimeout(() => {
      getList();
      this.setState({ loading: false, input: { name: "", price: 0 } });
    }, 1000);
  };

  handleOpenDialog = () => {
    const { isOpen } = this.state;
    this.setState({ isOpen: !isOpen });
  };

  handleOpenAdminDialog = () => {
    const { isAdminOpen } = this.state;
    this.setState({ isAdminOpen: !isAdminOpen });
  };

  render() {
    const { adminMode, loading: isLoading } = this.state;
    const { loading, items } = this.props;

    return (
      <React.Fragment>
        <Navbar className={Classes.DARK}>
          <Navbar.Group align={Alignment.LEFT}>
            <Navbar.Heading>Uncle's Store</Navbar.Heading>
            <Navbar.Divider />
          </Navbar.Group>
          <Navbar.Group align={Alignment.RIGHT}>
            <Button
              className={Classes.BUTTON}
              icon="cog"
              text="Admin Mode"
              intent="warning"
              onClick={() => this.handleOpenAdminDialog()}
            />
          </Navbar.Group>
        </Navbar>

        <div className={css.root}>
          {adminMode && (
            <div className={css.sidebar}>
              <h3>Admin Panel</h3>

              <Button
                intent="primary"
                className={Classes.LARGE}
                onClick={() => this.handleOpenDialog()}
              >
                Create an Item
              </Button>
            </div>
          )}

          <div className={css.body}>
            {items?.length < 1 && (
              <NonIdealState
                icon={this.state.icon ? "search" : undefined}
                title={adminMode ? "You have no items." : "We are restocking!"}
                description={
                  adminMode
                    ? "Add items from panel!"
                    : "Please wait for items to be added..."
                }
              />
            )}
            {loading || isLoading ? <Spinner /> : this.renderCards()}
          </div>
        </div>
        {this.renderDialog()}
        {this.renderAdminDialog()}
      </React.Fragment>
    );
  }
}

const mapToProps = ({ items, getList, loading }) => ({
  items,
  getList,
  loading,
});

export default connect(mapToProps, actions)(CatalogPage);
