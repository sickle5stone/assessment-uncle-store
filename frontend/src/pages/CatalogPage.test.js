import { fireEvent, render, screen, waitFor } from "@testing-library/react";

import CatalogPage from "./CatalogPage";
import { Provider } from "redux-zero/react";
import store from "../store/store";

test("renders no items message", () => {
  render(
    <Provider store={store}>
      <CatalogPage data={[]} />
    </Provider>
  );
  const noItems = screen.getByText(/Please wait for items to be added/i);
  expect(noItems).toBeInTheDocument();
});

test("render CatalogPage", () => {
  const { asFragment } = render(
    <Provider store={store}>
      <CatalogPage />{" "}
    </Provider>
  );

  expect(
    asFragment(
      <Provider store={store}>
        <CatalogPage />{" "}
      </Provider>
    )
  ).toMatchSnapshot();
});
