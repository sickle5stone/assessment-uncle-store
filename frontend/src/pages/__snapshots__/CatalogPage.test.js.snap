// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render CatalogPage 1`] = `
<DocumentFragment>
  <div
    class="bp3-navbar bp3-dark"
  >
    <div
      class="bp3-navbar-group bp3-align-left"
    >
      <div
        class="bp3-navbar-heading"
      >
        Uncle's Store
      </div>
      <div
        class="bp3-navbar-divider"
      />
    </div>
    <div
      class="bp3-navbar-group bp3-align-right"
    >
      <button
        class="bp3-button bp3-intent-warning bp3-button"
        type="button"
      >
        <span
          class="bp3-icon bp3-icon-cog"
          icon="cog"
        >
          <svg
            data-icon="cog"
            height="16"
            viewBox="0 0 16 16"
            width="16"
          >
            <desc>
              cog
            </desc>
            <path
              d="M15.19 6.39h-1.85c-.11-.37-.27-.71-.45-1.04l1.36-1.36c.31-.31.31-.82 0-1.13l-1.13-1.13a.803.803 0 00-1.13 0l-1.36 1.36c-.33-.17-.67-.33-1.04-.44V.79c0-.44-.36-.8-.8-.8h-1.6c-.44 0-.8.36-.8.8v1.86c-.39.12-.75.28-1.1.47l-1.3-1.3c-.3-.3-.79-.3-1.09 0L1.82 2.91c-.3.3-.3.79 0 1.09l1.3 1.3c-.2.34-.36.7-.48 1.09H.79c-.44 0-.8.36-.8.8v1.6c0 .44.36.8.8.8h1.85c.11.37.27.71.45 1.04l-1.36 1.36c-.31.31-.31.82 0 1.13l1.13 1.13c.31.31.82.31 1.13 0l1.36-1.36c.33.18.67.33 1.04.44v1.86c0 .44.36.8.8.8h1.6c.44 0 .8-.36.8-.8v-1.86c.39-.12.75-.28 1.1-.47l1.3 1.3c.3.3.79.3 1.09 0l1.09-1.09c.3-.3.3-.79 0-1.09l-1.3-1.3c.19-.35.36-.71.48-1.1h1.85c.44 0 .8-.36.8-.8v-1.6a.816.816 0 00-.81-.79zm-7.2 4.6c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
        <span
          class="bp3-button-text"
        >
          Admin Mode
        </span>
      </button>
    </div>
  </div>
  <div
    class="root"
  >
    <div
      class="body"
    >
      <div
        class="bp3-non-ideal-state"
      >
        <h4
          class="bp3-heading"
        >
          We are restocking!
        </h4>
        <div>
          Please wait for items to be added...
        </div>
      </div>
      <div
        class="bp3-spinner"
      >
        <div
          class="bp3-spinner-animation"
        >
          <svg
            height="50"
            stroke-width="8.00"
            viewBox="1.00 1.00 98.00 98.00"
            width="50"
          >
            <path
              class="bp3-spinner-track"
              d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
            />
            <path
              class="bp3-spinner-head"
              d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
              pathLength="280"
              stroke-dasharray="280 280"
              stroke-dashoffset="210"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
   
</DocumentFragment>
`;
